source/s3_compute_platform_info.c

```diff
@@ -0,0 +1,591 @@
+        if (!imds_client) {
+            goto tear_down;
+        }
+
+        aws_mutex_lock(&callback_info.mutex);
+        aws_imds_client_get_instance_info(imds_client, s_imds_client_on_get_instance_info_callback, &callback_info);
+        aws_condition_variable_wait_for_pred(
+            &callback_info.c_var, &callback_info.mutex, AWS_TIMESTAMP_SECS, s_completion_predicate, &callback_info);
+
+        aws_condition_variable_wait_pred(
+            &callback_info.c_var, &callback_info.mutex, s_client_shutdown_predicate, &callback_info);
+        aws_mutex_unlock(&callback_info.mutex);
+        aws_imds_client_release(imds_client);
+
+        if (callback_info.error_code) {
+            aws_raise_error(callback_info.error_code);
+            AWS_LOGF_ERROR(
+                AWS_LS_S3_CLIENT,
+                "id=%p: IMDS call failed with error %s.",
+                (void *)loader,
+                aws_error_debug_str(callback_info.error_code));
+        }
+
+        if (callback_info.instance_type) {
+            loader->lock_data.detected_instance_type = callback_info.instance_type;
+            loader->lock_data.current_env_platform_info.instance_type =
+                aws_byte_cursor_from_string(callback_info.instance_type);
+            s_add_platform_info_to_table(loader, &loader->lock_data.current_env_platform_info);
+            AWS_LOGF_INFO(
+                AWS_LS_S3_CLIENT,
+                "id=%p: Determined instance type to be %s, from IMDS. Caching.",
+                (void *)loader,
+                aws_string_bytes(loader->lock_data.detected_instance_type));
+        }
+
+    tear_down:
+        if (client_bootstrap) {
+            aws_client_bootstrap_release(client_bootstrap);
+        }
+
+        if (resolver) {
+            aws_host_resolver_release(resolver);
+        }
+
+        if (el_group) {
+            aws_event_loop_group_release(el_group);
+        }
+    }
+
+return_instance_and_unlock:
+    return_cur = loader->lock_data.current_env_platform_info.instance_type;
+    aws_mutex_unlock(&loader->lock_data.lock);
+
+    return return_cur;
+}
+
+const struct aws_s3_compute_platform_info *aws_s3_get_compute_platform_info_for_current_environment(
+    struct aws_s3_compute_platform_info_loader *loader) {
+    /* getting the instance type will set it on the loader the first time if it can */
+    aws_s3_get_ec2_instance_type(loader);
+    /* will never be mutated after the above call. */
+    return &loader->lock_data.current_env_platform_info;
+}
+
+const struct aws_s3_compute_platform_info *aws_s3_get_compute_platform_info_for_instance_type(
+    struct aws_s3_compute_platform_info_loader *loader,
+    struct aws_byte_cursor instance_type_name) {
+    aws_mutex_lock(&loader->lock_data.lock);
+    struct aws_hash_element *platform_info_element = NULL;
+    aws_hash_table_find(&loader->lock_data.compute_platform_info_table, &instance_type_name, &platform_info_element);
+    aws_mutex_unlock(&loader->lock_data.lock);
```

Comment:
🧐 annoying theoretical: someone calls this, gets a pointer, THEN calls for_current_environment(), which could mutate the data in the pointer they were holding.

I get that we don't want to do that IMDS call at library init

and I get that probably no one will ever call get_compute_platform_info_for_instance_type() except our tests

possible solutions that immediately come to mind:
1) this function also ensures we've done the IMDS lookup, so that everything is guaranteed const after we return (except as noted in another comment, we need to guarantee that lookup only happens once, by remembering failures as well as successes)
2) don't expose the for_instance_type() functions publicly, and document that they're not threadsafe

Reply:
Decided to go with 2 since it is simpler.